hw/xbox/mcpx_apu.c

```diff
@@ -473,6 +514,120 @@ static void ep_scratch_rw(void *opaque,
                       ptr, addr, len, dir);
 }
 
+static uint32_t circular_scatter_gather_rw(MCPXAPUState *d,
+                                           hwaddr sge_base,
+                                           unsigned int max_sge,
+                                           uint8_t *ptr,
+                                           uint32_t base, uint32_t end,
+                                           uint32_t cur,
+                                           size_t len,
+                                           bool dir)
+{
+    while (len > 0) {
```

Comment:
What if `len` >  `end-base`?

Reply:
I don't remember if I tested this on hardware, but this case does also not occur in the DirectSound code.

I assume the behaviour to be as it is here - and we can fix it later if this assumption is wrong.

Reply:
Perhaps we should `assert` to catch instances when our assumption here is wrong.

Reply:
If need-be we can just test this on hardware and document it.

But as we have unit testing for this anyway, it's not critical to do before a merge either. Detecting problems here, including debugging this code, is very easy. I have also confirmed this does not occur in DirectSound.

Even *if* this code does not work, we'll be able to *hear* that it's wrong. So errors in this code wouldn't go unnoticed.